### ISSUE TITLE
Adaptative typing bug

### ISSUE THREAD

**Issue by R4PaSs** [privat/nit#1439]
Hi, it seems there's a little bug in the adaptative type system.

I have this sample code:

~~~nit
class A
end

class B
        super A
end

class C
        super A
end

class E
        super B

        var c: A
end

class D
        fun foo(s: B) do
                var c: A = s
                if not c isa E then abort
                loop
                        if c isa C then return
                        if not c isa E then abort
                        c = c.c
                end
        end
end
~~~

When trying to compile, I get this warning: `Warning: expression is already a `E`. (useless-type-test)`

I thought, oh well fine, wow adaptative typing is so intelligent it could figure that `c` could only be a E since it was not a C, how beautiful.

But, I was wrong.

When I remove the type test, here's what I get: `Error: method c does not exists in A`

Hm, damn. This is embarassing.

**Comment by privat** [comment]
As explained in #1257 warnings are still more or less displayed while type adaptation is working, so based on insufficient information. The initial warning should have not been displayed.

This is still a bug to work on.

**Comment by privat** [comment]
closed by #1524
